### PR TITLE
Seperate the MMU and cache from the lsu

### PR DESCRIPTION
**Pull request by h0lyalg0rithm**
I kept the logic the same as it was earlier.
One of the design considerations was to use the BIU to communicate with the cache and MMU.However it was later decided to keep the communication simple until we design a more robust MMU design

### CODE REVIEW

**Review by knute-sifive — APPROVED** [review]
LGTM

File: core/LSU.cpp [hunk 1]
```diff
@@ -257,7 +263,7 @@ namespace olympia
         }
 
         // Access TLB, and check TLB hit or miss
-        bool TLB_HIT = MMULookup_(mem_access_info_ptr);
+        bool TLB_HIT = mmu_->Lookup(mem_access_info_ptr);
```
Comment (knute-sifive) [inline core/LSU.cpp:266]: Two things:
1. While we haven't really established coding style guidelines, this variable name should not be in all caps.  Request `tlb_hit` to match the other code
2. This variable can be `const` : `const bool tlb_hit =`.  This is helpful is avoiding accidental assignments when you intended to compare.

File: core/LSU.cpp [hunk 2]
```diff
@@ -353,22 +359,22 @@ namespace olympia
 
         if (cacheBypass) {
```
Comment (knute-sifive) [inline core/LSU.cpp:360]: Can you change the variable name to be consistent? `cache_bypass`

File: core/LSU.cpp [hunk 3]
```diff
@@ -353,22 +359,22 @@ namespace olympia
...
             if (isAlreadyHIT) {
```
Comment (knute-sifive) [inline core/LSU.cpp:361]: Same here.

File: core/LSU.cpp [hunk 4]
```diff
@@ -353,22 +359,22 @@ namespace olympia
...
             else {
-                sparta_assert(false, "Cache access is bypassed without a valid reason!");
+                sparta_assert(false, "DCache access is bypassed without a valid reason!");
             }
             return;
         }
 
         // Access cache, and check cache hit or miss
-        const bool CACHE_HIT = cacheLookup_(mem_access_info_ptr);
+        const bool CACHE_HIT = data_cache_->cacheLookup(mem_access_info_ptr);
```
Comment (knute-sifive) [inline core/LSU.cpp:377]: Variable name change, please.

File: core/LSU.cpp [hunk 5]
```diff
@@ -596,6 +608,36 @@ namespace olympia
...
+    void LSU::popLoadQueue_(const InstPtr & inst_ptr)
+    {
+        // Look for the instruction to be completed, and remove it from load queue
+        for (auto iter = load_queue_.begin(); iter != load_queue_.end(); iter++) {
+            if ((*iter)->getInstPtr() == inst_ptr) {
+                load_queue_.erase(iter);
+
+                return;
+            }
+        }
```
Comment (knute-sifive) [inline core/LSU.cpp:621]: A faster way to do this: when the load is allocated into the load queue, store the index (iterator) of its position in the request.  Then, when popping, check that iterator's validity and if valid, just erase it.

File: core/LSU.cpp [hunk 6]
```diff
@@ -596,6 +608,36 @@ namespace olympia
...
+    void LSU::popStoreQueue_(const InstPtr & inst_ptr)
+    {
+        // Look for the instruction to be completed, and remove it from store queue
+        for (auto iter = store_queue_.begin(); iter != store_queue_.end(); iter++) {
+            if ((*iter)->getInstPtr() == inst_ptr) {
+                store_queue_.erase(iter);
+
+                return;
+            }
+        }
```
Comment (knute-sifive) [inline core/LSU.cpp:636]: Same suggestion here as above.

File: core/LSU.cpp [hunk 7]
```diff
@@ -643,43 +685,28 @@ namespace olympia
...
+    void LSU::appendLoadQueue_(const LoadStoreInstInfoPtr & inst_info_ptr)
     {
-        const InstPtr & inst_ptr = mem_access_info_ptr->getInstPtr();
-        uint64_t vaddr = inst_ptr->getTargetVAddr();
+        sparta_assert(load_queue_.size() <= ld_queue_size_,
+                      "Appending load queue causes overflows!");
 
-        bool tlb_hit = false;
+        // Always append newly dispatched instructions to the back of load queue
+        load_queue_.push_back(inst_info_ptr);
```
Comment (knute-sifive) [inline core/LSU.cpp:695]: Yeah, you can do this:
```suggestion
        inst_info_ptr->storeLQPosition(load_queue_.push_back(inst_info_ptr));
```

File: core/LSU_MMU_Shared.hpp
Comment (knute-sifive) [inline core/LSU_MMU_Shared.hpp:1]: Why is called "shared"?

File: core/LSU_MMU_Shared.hpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            return mmu_access_state_;
+        }
+
+        void setMMUState(const MMUState &state) {
+            mmu_access_state_ = state;
+        }
+
+        const CacheState &getCacheState() const {
+            return cache_access_state_;
+        }
```
Comment (knute-sifive) [inline core/LSU_MMU_Shared.hpp:70]: MMUState and CacheState are enums (POD types).  Returning a reference converts them to a pointer and de dereferenced.  Makes sense to just return by value.   (It's a good chance the compiler will do that for you)

File: core/MMU.cpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    bool MMU::Lookup(const MemoryAccessInfoPtr &mem_access_info_ptr) {
+        const InstPtr &inst_ptr = mem_access_info_ptr->getInstPtr();
+        uint64_t vaddr = inst_ptr->getTargetVAddr();
+
+        bool tlb_hit = false;
+
+        // C++ comma operator: assign tlb_hit first, then evaluate it. Just For Fun
+        if (tlb_hit = tlb_always_hit_, tlb_hit) {
+        } else {
```
Comment (knute-sifive) [inline core/MMU.cpp:24]: Might as well convert to 
```suggestion
        bool tlb_hit = tlb_always_hit_;
        if(false == tlb_hit) {
             // miss code
```

File: core/MMU.cpp [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+            }
+        }
+
+
+        if (tlb_always_hit_) {
+            ILOG("TLB HIT all the time: vaddr=0x" << std::hex << vaddr);
+            tlb_hits_++;
+        } else if (tlb_hit) {
+            ILOG("TLB HIT: vaddr=0x" << std::hex << vaddr);
+            tlb_hits_++;
```
[the remark this answers is not in the document]
Comment (h0lyalg0rithm) [reply core/MMU.cpp:40]: I think this was more for debugging purposes

File: core/LSU_MMU_Shared.hpp
Comment (h0lyalg0rithm) [reply core/LSU_MMU_Shared.hpp:1]: I though I might have to put in more structures apart from the memoryaccessinfo.
But it makes sense to rename it to `MemoryAccessInfo.hpp`

File: core/LSU.cpp (hunk 7 above)
Comment (h0lyalg0rithm) [reply core/LSU.cpp:696]: I will remove the code from this PR to keep the changes related to the separation of the cache and mmu

**Comment by h0lyalg0rithm** [comment]
@knute-sifive Both Vineeth and I have completed this PR.We would use the final commit to branch out and work on our changes to the lsu and cache

File: core/LSU.cpp [hunk 11]
```diff
@@ -925,6 +612,64 @@ namespace olympia
...
+        // Check if flushing event occurred just now
+        if (!ldst_pipeline_.isValid(stage_id)) {
+            return;
+        }
+
+        const MemoryAccessInfoPtr & mem_access_info_ptr = ldst_pipeline_[stage_id];
+        ILOG(mem_access_info_ptr);
+
+        bool isAlreadyHIT = (mem_access_info_ptr->getMMUState() == MemoryAccessInfo::MMUState::HIT);
+        bool MMUBypass = isAlreadyHIT;
```
Comment (knute-sifive) [inline core/LSU.cpp:632]: Duplicate `bool` variables.  Remove one; make the other `const` and change the variable name to match the rest of the code

File: core/LSU.cpp [hunk 12]
```diff
@@ -925,6 +612,64 @@ namespace olympia
...
+
+    // Handle MMU access request
+    void LSU::handleMMULookupReq2_()
+    {
+        if(!mmu_hit_)
+            ldst_pipeline_.invalidateStage(static_cast<uint32_t>(PipelineStage::MMU_LOOKUP));
+    }
+
+    void LSU::getInstFromMMU_(const MemoryAccessInfoPtr &memory_access_info_ptr) {
+        auto inst_ptr = memory_access_info_ptr->getInstPtr();
```
Comment (knute-sifive) [inline core/LSU.cpp:652]: `const auto &` to avoid the copy

**Pull request closed without merging**